src/bootstrap/builder.rs

```diff
@@ -616,32 +630,29 @@ impl<'a> Builder<'a> {
                 install::Rustc
             ),
             Kind::Run => describe!(run::ExpandYamlAnchors, run::BuildManifest, run::BumpStage0),
+            // These subcommands are special-cased in Build::build() and don't have typical steps
+            Kind::Format | Kind::Clean | Kind::Setup => vec![],
         }
     }
 
-    pub fn get_help(build: &Build, subcommand: &str) -> Option<String> {
-        let kind = match subcommand {
-            "build" | "b" => Kind::Build,
-            "doc" | "d" => Kind::Doc,
-            "test" | "t" => Kind::Test,
-            "bench" => Kind::Bench,
-            "dist" => Kind::Dist,
-            "install" => Kind::Install,
-            _ => return None,
-        };
+    pub fn get_help(build: &Build, kind: Kind) -> Option<String> {
+        let step_descriptions = Builder::get_step_descriptions(kind);
+        if step_descriptions.is_empty() {
+            return None;
```

Comment:
This was part of the logic for printing the hint to run with `-h -v` to see available paths if and only if there are paths that we could list for that subcommand.

For example, in Rust now if you run `./x.py clippy -h` it will suggest `Run './x.py clippy -h -v' to see a list of available paths.`, but when you run `./x.py clippy -v -v` it doesn't actually list any paths.

My assumption was that all of these commands could have useful paths to pass to them but I might be mistaken about that. It seems like check and clippy would take paths just like build does, but it kinda seems like clippy runs for "everything" in the build graph anyway and passing paths doesn't do much.

Is this specific list of commands which could print a list of paths for help text intentional and the others (clippy, check, fix) are supposed to be excluded?

Reply:
> My assumption was that all of these commands could have useful paths to pass to them but I might be mistaken about that.

This is not currently true. There are several related issues:
1. Not all commands accept paths. `x clippy`, as you noticed, runs on the whole tree. `x clean` does the same.
2. Some commands that do not accept paths *do not check whether a path was passed* - `x clippy src/bootstrap` is exactly the same as an unconditional `x clippy`.
3. Some commands the *do* accept paths do not go through `StepDescription`. For example `x fmt src/bootstrap` is perfectly valid and limits the changes to bootstrap itself, but there is no `impl Step for Format` description, because `Step` has no way to say "all paths".

Fixing those issues should wait for follow up PRs, I think. In the meantime, can you do the following:
- [ ] Remove `clippy` and `fix` from the `check` arm and down to the `vec![]` arm. `clippy` is certainly not supported, and `fix` has been broken for ages.
- [ ] Change the code in `usage` that calls `get_help` to panic if `-h -v` is called for a subcommand that does not support paths. We should give the user an idea that something is broken rather than silently ignoring the flag.

Reply:
I understand now why this changed and the logic here looks good to me. See my comment below for next steps.